File: web/src/lib/schedule.ts
/**
 * Jadval mantiqi — React'dan mustaqil sof funksiyalar.
 * Python tomondagi `bot/services/schedule.py` bilan bir xil algoritm.
 */
import type {
  ActiveBus,
  Departure,
  Direction,
  Route,
  TimeString,
  Trip,
} from "@/types/route";
import { minutesOfDay, parseTime } from "@/lib/time";

/** Uchrashuv chegarasi — chiziq uzunligining ulushi. Bot bilan bir xil. */
export const MEETING_THRESHOLD_RATIO = 0.06;

export const DIRECTIONS: readonly Direction[] = ["shaharga", "qishloqqa"] as const;

export const DIRECTION_LABEL: Record<Direction, string> = {
  shaharga: "Shaharga",
  qishloqqa: "Qishloqqa",
};

export function oppositeOf(direction: Direction): Direction {
  return direction === "shaharga" ? "qishloqqa" : "shaharga";
}

/** Reysdagi `null` bo'lmagan vaqtlar, bekat tartibida. */
function statedTimes(trip: Trip): TimeString[] {
  return trip.times.filter((time): time is TimeString => time !== null);
}

/**
 * Reys yo'nalishi. `stops` qishloqdan shaharga tartibda yozilgani uchun
 * vaqt index bo'yicha oshsa — shaharga, kamaysa — qishloqqa.
 * Aniqlab bo'lmasa (2 tadan kam vaqt) — null.
 */
export function tripDirection(trip: Trip): Direction | null {
  const stated = statedTimes(trip);
  if (stated.length < 2) return null;
  return stated[stated.length - 1]! > stated[0]! ? "shaharga" : "qishloqqa";
}

/**
 * Reysning uzluksiz pozitsiyasi (0 .. bekatlar_soni−1).
 * Hali boshlanmagan yoki tugagan bo'lsa — null.
 * Bitta formula ikkala yo'nalish uchun ham to'g'ri, chunki interpolyatsiya
 * vaqt oralig'i bo'yicha bajariladi.
 */
export function tripPosition(trip: Trip, now: Date): number | null {
  const current = minutesOfDay(now);

  for (let i = 0; i < trip.times.length - 1; i += 1) {
    const first = trip.times[i];
    const second = trip.times[i + 1];
    if (first == null || second == null) continue;

    const start = parseTime(first);
    const end = parseTime(second);
    if (start === end) continue;

    const low = Math.min(start, end);
    const high = Math.max(start, end);
    if (current < low || current > high) continue;

    return i + (current - start) / (end - start);
  }

  return null;
}

/** Qarama-qarshi yo'nalishdagi yaqin avtobuslarni juftlab belgilaydi. */
export function markMeetings(
  buses: readonly Omit<ActiveBus, "meetingWith">[],
  stopCount: number,
): ActiveBus[] {
  const threshold = (stopCount - 1) * MEETING_THRESHOLD_RATIO;
  const partners = new Map<string, string>();

  for (let i = 0; i < buses.length; i += 1) {
    for (let j = i + 1; j < buses.length; j += 1) {
      const a = buses[i]!;
      const b = buses[j]!;
      if (a.direction === b.direction) continue;
      if (Math.abs(a.position - b.position) <= threshold) {
        partners.set(a.busId, b.busId);
        partners.set(b.busId, a.busId);
      }
    }
  }

  return buses.map((bus) => ({ ...bus, meetingWith: partners.get(bus.busId) ?? null }));
}

/** Hozir yo'lda bo'lgan barcha avtobuslar. */
export function activeBuses(route: Route, now: Date): ActiveBus[] {
  const moving: Omit<ActiveBus, "meetingWith">[] = [];

  for (const trip of route.trips) {
    const direction = tripDirection(trip);
    const position = tripPosition(trip, now);
    if (direction === null || position === null) continue;
    moving.push({ busId: trip.bus_id, direction, position });
  }

  return markMeetings(moving, route.stops.length);
}

/** Bekatdagi barcha jo'nashlar, yo'nalish bo'yicha ajratilgan va tartiblangan. */
export function departuresAt(
  route: Route,
  stopIndex: number,
): Record<Direction, TimeString[]> {
  const result: Record<Direction, TimeString[]> = { shaharga: [], qishloqqa: [] };
  if (stopIndex < 0 || stopIndex >= route.stops.length) return result;

  for (const trip of route.trips) {
    const direction = tripDirection(trip);
    const time = trip.times[stopIndex];
    if (direction === null || time == null) continue;
    result[direction].push(time);
  }

  for (const times of Object.values(result)) {
    times.sort();
  }
  return result;
}

/** Bekatdagi keyingi jo'nashlar (har yo'nalishdan `limit` tadan). */
export function nextDeparturesAt(
  route: Route,
  stopIndex: number,
  now: Date,
  limit = 3,
): Record<Direction, Departure[]> {
  const current = minutesOfDay(now);
  const all = departuresAt(route, stopIndex);
  const result: Record<Direction, Departure[]> = { shaharga: [], qishloqqa: [] };

  for (const direction of DIRECTIONS) {
    result[direction] = all[direction]
      .map((time) => ({ time, direction, minutesAway: parseTime(time) - current }))
      .filter((departure) => departure.minutesAway >= 0)
      .slice(0, limit);
  }

  return result;
}

/** Bir yo'nalish bo'yicha bugungi birinchi reys (kun tugaganda ko'rsatiladi). */
export function firstDepartureOfDay(
  route: Route,
  stopIndex: number,
  direction: Direction,
): TimeString | null {
  return departuresAt(route, stopIndex)[direction][0] ?? null;
}

File: bot/services/schedule.py
"""Jadval hisob-kitoblari — Telegram'dan mustaqil, sof funksiyalar.

Bu modul `bot/` ichida ham, kelajakdagi API/admin panelda ham qayta
ishlatiladi. Shu sabab bu yerda aiogram importlari YO'Q.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, time

from bot.models import Direction, Route, TimeStr, Trip, parse_time

# Ikki avtobus "uchrashdi" deb hisoblanadigan masofa —
# chiziq uzunligining ulushi. Frontend (web/src/lib/schedule.ts) bilan bir xil.
MEETING_THRESHOLD_RATIO = 0.06


def _now_minutes(now: datetime | time) -> int:
    return now.hour * 60 + now.minute


def departures_at(route: Route, stop_index: int) -> dict[Direction, list[TimeStr]]:
    """Bekatdagi barcha reys vaqtlarini yo'nalish bo'yicha ajratadi.

    Har bir ro'yxat o'sish tartibida qaytariladi.
    """
    if not 0 <= stop_index < route.stop_count:
        raise IndexError(f"Bekat indeksi chegaradan tashqarida: {stop_index}")

    result: dict[Direction, list[TimeStr]] = {d: [] for d in Direction}
    for trip in route.trips:
        direction = trip.direction
        departure = trip.time_at(stop_index)
        if direction is None or departure is None:
            continue
        result[direction].append(departure)

    for times in result.values():
        times.sort()
    return result


def next_departures(times: list[TimeStr], now: datetime, limit: int = 3) -> list[TimeStr]:
    """Hozirgi vaqtdan keyingi eng yaqin `limit` ta jo'nashni qaytaradi.

    Kunlik reyslar tugagan bo'lsa — bo'sh ro'yxat.
    """
    current = _now_minutes(now)
    return [t for t in times if parse_time(t) >= current][:limit]


def minutes_until(departure: TimeStr, now: datetime) -> int:
    """Jo'nashgacha necha daqiqa qolgani (o'tib ketgan bo'lsa — manfiy)."""
    return parse_time(departure) - _now_minutes(now)


def trip_position(trip: Trip, now: datetime) -> float | None:
    """Reysning hozirgi uzluksiz pozitsiyasi (0 .. bekatlar_soni-1).

    Reys hali boshlanmagan yoki allaqachon tugagan bo'lsa — None.
    Ikkala yo'nalish uchun ham bitta formula ishlaydi, chunki interpolyatsiya
    vaqt oralig'i bo'yicha, bekat indeksi bo'yicha emas.
    """
    current = _now_minutes(now)
    times = trip.times

    for i in range(len(times) - 1):
        first, second = times[i], times[i + 1]
        if first is None or second is None:
            continue

        start, end = parse_time(first), parse_time(second)
        low, high = min(start, end), max(start, end)
        if not (low <= current <= high) or start == end:
            continue

        progress = (current - start) / (end - start)
        return i + progress

    return None


@dataclass(frozen=True, slots=True)
class ActiveBus:
    """Ayni damda yo'lda bo'lgan avtobus."""

    bus_id: str
    direction: Direction
    position: float
    meeting_with: str | None = None

    @property
    def is_meeting(self) -> bool:
        return self.meeting_with is not None


def active_buses(route: Route, now: datetime) -> list[ActiveBus]:
    """Hozir harakatdagi barcha avtobuslar, uchrashuvlari belgilangan holda."""
    buses: list[ActiveBus] = []
    for trip in route.trips:
        position = trip_position(trip, now)
        if position is None or trip.direction is None:
            continue
        buses.append(ActiveBus(bus_id=trip.bus_id, direction=trip.direction, position=position))

    return find_meetings(buses, route.stop_count)


def find_meetings(buses: list[ActiveBus], stop_count: int) -> list[ActiveBus]:
    """Qarama-qarshi yo'nalishdagi yaqin avtobuslarni juftlab belgilaydi."""
    threshold = (stop_count - 1) * MEETING_THRESHOLD_RATIO
    partners: dict[str, str] = {}

    for i, first in enumerate(buses):
        for second in buses[i + 1:]:
            if first.direction is second.direction:
                continue
            if abs(first.position - second.position) <= threshold:
                partners[first.bus_id] = second.bus_id
                partners[second.bus_id] = first.bus_id

    return [
        ActiveBus(
            bus_id=bus.bus_id,
            direction=bus.direction,
            position=bus.position,
            meeting_with=partners.get(bus.bus_id),
        )
        for bus in buses
    ]
